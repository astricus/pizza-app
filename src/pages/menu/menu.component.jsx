import React, { useEffect, lazy, Suspense } from 'react';
import { Route } from 'react-router-dom';
import { connect } from 'react-redux';

import { fetchCollectionsStart } from '../../redux/menu/menu.actions';

import Spinner from '../../components/spinner/spinner.component';

import { MenuPageContainer } from './menu.styles';

const CollectionPageContainer = lazy(() =>
  import('../collection/collection.container')
);

export const MenuPage = ({ fetchCollectionsStart, match }) => {
  useEffect(() => {
    fetchCollectionsStart();
  }, [fetchCollectionsStart]);

  return (
    <MenuPageContainer>
      <Suspense fallback={<Spinner />}>
        <Route
          exact
          path={`${match.path}`}
          component={CollectionPageContainer}
        />
      </Suspense>
    </MenuPageContainer>
  );
};

const mapDispatchToProps = (dispatch) => ({
  fetchCollectionsStart: () => dispatch(fetchCollectionsStart()),
});

export default connect(null, mapDispatchToProps)(MenuPage);
