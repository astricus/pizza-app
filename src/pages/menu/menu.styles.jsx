import styled from 'styled-components';

export const MenuPageContainer = styled.div`
  width: 100%;
`;
