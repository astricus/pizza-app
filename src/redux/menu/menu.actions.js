import MenuActionTypes from './menu.types';

export const fetchCollectionsStart = () => ({
  type: MenuActionTypes.FETCH_COLLECTIONS_START,
});

export const fetchCollectionsSuccess = (collectionsMap) => ({
  type: MenuActionTypes.FETCH_COLLECTIONS_SUCCESS,
  payload: collectionsMap,
});

export const fetchCollectionsFailure = (errorMessage) => ({
  type: MenuActionTypes.FETCH_COLLECTIONS_FAILURE,
  payload: errorMessage,
});

export const setCurrency = (currency) => ({
  type: MenuActionTypes.SET_CURRENCY,
  payload: currency,
});
