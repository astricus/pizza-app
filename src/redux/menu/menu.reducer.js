import MenuActionTypes from './menu.types';

const INITIAL_STATE = {
  collections: null,
  isFetching: false,
  currencyCoefficient: 1,
  currencySymbol: '$',
  errorMessage: undefined,
};

const menuReducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case MenuActionTypes.FETCH_COLLECTIONS_START:
      return {
        ...state,
        isFetching: true,
      };
    case MenuActionTypes.FETCH_COLLECTIONS_SUCCESS:
      return {
        ...state,
        isFetching: false,
        collections: action.payload,
      };
    case MenuActionTypes.FETCH_COLLECTIONS_FAILURE:
      return {
        ...state,
        isFetching: false,
        errorMessage: action.payload,
      };
    case MenuActionTypes.SET_CURRENCY:
      return {
        ...state,
        currencyCoefficient: action.payload === 'EUR' ? 0.92 : 1,
        currencySymbol: action.payload === 'EUR' ? 'â‚¬' : '$',
      };
    default:
      return state;
  }
};

export default menuReducer;
