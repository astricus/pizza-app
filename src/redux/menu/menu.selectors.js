import { createSelector } from 'reselect';

const selectMenu = (state) => state.menu;

export const selectCollections = createSelector(
  [selectMenu],
  (menu) => menu.collections
);

export const selectCollectionsForPreview = createSelector(
  [selectCollections],
  (collections) =>
    collections ? Object.keys(collections).map((key) => collections[key]) : []
);

export const selectCurrencyCoefficient = createSelector(
  [selectMenu],
  (menu) => menu.currencyCoefficient
);

export const selectCurrencySymbol = createSelector(
  [selectMenu],
  (menu) => menu.currencySymbol
);

export const selectCollection = (collectionUrlParam) =>
  createSelector([selectCollections], (collections) =>
    collections ? collections[collectionUrlParam] : null
  );

export const selectIsCollectionFetching = createSelector(
  [selectMenu],
  (menu) => menu.isFetching
);

export const selectIsCollectionsLoaded = createSelector(
  [selectMenu],
  (menu) => !!menu.collections
);
