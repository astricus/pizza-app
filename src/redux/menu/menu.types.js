const MenuActionTypes = {
  FETCH_COLLECTIONS_START: 'FETCH_COLLECTIONS_START',
  FETCH_COLLECTIONS_SUCCESS: 'FETCH_COLLECTIONS_SUCCESS',
  FETCH_COLLECTIONS_FAILURE: 'FETCH_COLLECTIONS_FAILURE',
  SET_CURRENCY: 'SET_CURRENCY',
};

export default MenuActionTypes;
