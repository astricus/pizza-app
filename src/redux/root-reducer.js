import { combineReducers } from 'redux';
import { persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import userReducer from './user/user.reducer';
import cartReducer from './cart/cart.reducer';
import orderReducer from './order/order.reducer';
import menuReducer from './menu/menu.reducer';

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['cart'],
};

const rootReducer = combineReducers({
  user: userReducer,
  order: orderReducer,
  cart: cartReducer,
  menu: menuReducer,
});

export default persistReducer(persistConfig, rootReducer);
