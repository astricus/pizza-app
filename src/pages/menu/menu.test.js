import React from 'react';
import { mount } from 'enzyme';
import { combineReducers, createStore } from 'redux';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { MenuPage } from './menu.component';

export const createMockStore = ({ state, reducers }) => {
  const store = createStore(combineReducers(reducers), state);
  return {
    ...store,
    persistor: {
      persist: () => null,
    },
  };
};

describe('MenuPage', () => {
  let wrapper;
  let mockFetchCollectionsStart;
  let store;

  beforeEach(() => {
    const mockReducer = (
      state = {
        isFetching: true,
      },
      action
    ) => state;

    const mockState = {
      menu: {
        isFetching: true,
      },
    };

    mockFetchCollectionsStart = jest.fn();

    store = createMockStore({
      state: mockState,
      reducers: { menu: mockReducer },
    });

    const mockMatch = {
      path: '',
    };

    const mockProps = {
      match: mockMatch,
      fetchCollectionsStart: mockFetchCollectionsStart,
    };

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <MenuPage {...mockProps} />
        </Provider>
      </BrowserRouter>
    );
  });

  it('should render MenuPage component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render MenuPage component', () => {
    expect(mockFetchCollectionsStart).toHaveBeenCalled();
  });
});
