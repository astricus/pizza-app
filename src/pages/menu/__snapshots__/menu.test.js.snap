// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuPage should render MenuPage component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "dispatch": [Function],
          "getState": [Function],
          "persistor": Object {
            "persist": [Function],
          },
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        }
      }
    >
      <MenuPage
        fetchCollectionsStart={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        match={
          Object {
            "path": "",
          }
        }
      >
        <styled.div>
          <div
            className="sc-AxiKw gtPwVX"
          >
            <Suspense
              fallback={<Spinner />}
            >
              <Route
                component={
                  Object {
                    "$$typeof": Symbol(react.lazy),
                    "_ctor": [Function],
                    "_result": Object {
                      "$$typeof": Symbol(react.memo),
                      "WrappedComponent": [Function],
                      "compare": null,
                      "displayName": "Connect(Component)",
                      "type": [Function],
                    },
                    "_status": 1,
                  }
                }
                exact={true}
                path=""
              />
              <Spinner>
                <styled.div>
                  <div
                    className="sc-AxjAm HvuBM"
                  >
                    <styled.div>
                      <div
                        className="sc-AxirZ hCASNt"
                      />
                    </styled.div>
                  </div>
                </styled.div>
              </Spinner>
            </Suspense>
          </div>
        </styled.div>
      </MenuPage>
    </Provider>
  </Router>
</BrowserRouter>
`;
